### PR TITLE
Print all output in nrepl client

### PR DESCRIPTION
**Pull request by SevereOverfl0w**
Before submitting a PR make sure the following things have been done:

- [x] The commits are consistent with our [contribution guidelines](CONTRIBUTING.md)
- [ ] You've added tests to cover your change(s)
- [x] All tests are passing
- [x] The new code is not generating reflection warnings
- [x] You've updated the [changelog](../CHANGELOG.md)(that only applies to user-visible changes)

Thanks!

### CODE REVIEW

**Comment by alexander-yakushev** [comment]
Hi, thanks for the fix! Overall, it seems reasonable. Do you think it is possible to combine the two `threading/run-with` into one? Just have a thread that prints stuff that arrives. Doesn't even have to look at the ID, there can be theoretically messages arriving that are not tied to any ID.

I'm also wondering what to do with the races. It seems like the `out` and the `value` can be printed in different order each time.

**Comment by SevereOverfl0w** [comment]
I was pondering this yesterday as well.  I think we can't combine the two, because other clients could be issuing commands which produce output which shouldn't be mirrored to the command line client.  That `nil` id check is a bit of a hack for startup.

Regarding the racing, I was thinking that maybe instead of doing a `doseq` over the responses, instead everything is handled by the thread I start.  The tricky bit is that the REPL loop needs a signal that the message is "done" being printed.  That's a bit more difficult but should be solid in terms of races.

I'll have a noodle at that now.

**Comment by alexander-yakushev** [comment]
> I was pondering this yesterday as well. I think we can't combine the two, because other clients could be issuing commands which produce output which shouldn't be mirrored to the command line client. That nil id check is a bit of a hack for startup.

I was under the impression that each `nrepl.core/client` will only see the messages from its own communication. Isn't it the case, @bbatsov? On the other hand, there doesn't seem to be anything but the message `id` to distinguish such things.

*CONT:*

I've just checked by starting to cmdline clients for a single server, and their output (and the messages received from the `(client)` don't intermingle. I suppose that's because each session gets its own `*out*` that forwards the output only to the session's own transport.

So, I think we should be quite safe listening and reacting to all messages on the client. What do you think, @bbatsov?

**Comment by alexander-yakushev** [comment]
> Regarding the racing, I was thinking that maybe instead of doing a doseq over the responses, instead everything is handled by the thread I start.

~You wouldn't be able to do that directly because at some point the thread has to block on listening to user's input (the "read") but at the same time be ready to accept and display the asynchronously arriving print messages. It's possible to do that somehow with NIO selectors or something, but that would mean overcomplicating a quite simple function.~

You'll still need two threads but one will only `(read)` and send the messages to the server, and the other will dispatch all messages and communicate the changed `*ns*` back to the first one (it's probably the only reason they have to talk).

**Comment by SevereOverfl0w** [comment]
Great, that simplifies things then.  I'll push something up now which does the coordinating threads which I think should be comprehensive in terms of printing _everything_ and coordinating the REPL loop with the thread.

File: src/clojure/nrepl/cmdline.clj [hunk 1]
```diff
@@ -93,16 +95,22 @@ Exit:      Control+D or (exit) or (quit)"
...
-      ;; Doesn't matter which executor we hijack here, just don't use `future`.
-      (->> (client)
-           (take-while #(nil? (:id %)))
-           (run! #(when-let [msg (:out %)] (print msg)))))
-    (Thread/sleep 50)
+      (run!
+       (fn [res]
+         (some-> (:out res) out)
+         (some-> (:err res) err)
+         (when-let [p (and (seq (set/intersection #{:done "done"} (set (:status res))))
```
Comment (alexander-yakushev) [inline src/clojure/nrepl/cmdline.clj:106]: Something like `(some #{:done "done} (:status res))` will be more compact.

File: src/clojure/nrepl/cmdline.clj [hunk 2]
```diff
@@ -93,16 +95,22 @@ Exit:      Control+D or (exit) or (quit)"
...
     (println (repl-intro))
-    ;; We take 50ms to listen to any greeting messages, and display the value
-    ;; in the `:out` slot.
     (threading/run-with @threading/transport-executor
-      ;; Doesn't matter which executor we hijack here, just don't use `future`.
-      (->> (client)
-           (take-while #(nil? (:id %)))
-           (run! #(when-let [msg (:out %)] (print msg)))))
-    (Thread/sleep 50)
+      (run!
```
Comment (alexander-yakushev) [inline src/clojure/nrepl/cmdline.clj:102]: `doseq` would save a few LOC over `run!` but that's up to you.

File: src/clojure/nrepl/cmdline.clj [hunk 3]
```diff
@@ -93,16 +95,22 @@ Exit:      Control+D or (exit) or (quit)"
...
                      out    print
                      value  println}}]
-  (let [client (nrepl/client transport Long/MAX_VALUE)]
+  (let [client (nrepl/client transport Long/MAX_VALUE)
+        awaiting (atom {})]
     (println (repl-intro))
-    ;; We take 50ms to listen to any greeting messages, and display the value
-    ;; in the `:out` slot.
     (threading/run-with @threading/transport-executor
-      ;; Doesn't matter which executor we hijack here, just don't use `future`.
```
Comment (alexander-yakushev) [inline src/clojure/nrepl/cmdline.clj:101]: I would prefer to preserve this comment.

**Review by alexander-yakushev — COMMENTED** [review]
Looks good to me, left a few nits.

File: src/clojure/nrepl/cmdline.clj [hunk 4]
```diff
@@ -93,16 +95,22 @@ Exit:      Control+D or (exit) or (quit)"
...
-    (Thread/sleep 50)
+      (run!
+       (fn [res]
+         (some-> (:out res) out)
+         (some-> (:err res) err)
+         (when-let [p (and (seq (set/intersection #{:done "done"} (set (:status res))))
+                           (get @awaiting (:id res)))]
+           (deliver p true))
+         (when (some-> (:id res)
+                       (str/starts-with? "nrepl.cmdline-"))
```
Comment (alexander-yakushev) [inline src/clojure/nrepl/cmdline.clj:110]: No `starts-with?` in Clojure 1.7, gotta use interop.

File: src/clojure/nrepl/cmdline.clj [hunk 5]
```diff
@@ -93,16 +94,21 @@ Exit:      Control+D or (exit) or (quit)"
...
+        awaiting (atom {})]
     (println (repl-intro))
-    ;; We take 50ms to listen to any greeting messages, and display the value
-    ;; in the `:out` slot.
     (threading/run-with @threading/transport-executor
       ;; Doesn't matter which executor we hijack here, just don't use `future`.
-      (->> (client)
-           (take-while #(nil? (:id %)))
-           (run! #(when-let [msg (:out %)] (print msg)))))
-    (Thread/sleep 50)
```
Comment (alexander-yakushev) [inline src/clojure/nrepl/cmdline.clj:105]: I think this is still needed. We want the greeting to be printed before the `user=>` prompt, this delay increases the chances of it happening.

File: src/clojure/nrepl/cmdline.clj [hunk 6]
```diff
@@ -93,16 +94,21 @@ Exit:      Control+D or (exit) or (quit)"
                      err    print
                      out    print
                      value  println}}]
-  (let [client (nrepl/client transport Long/MAX_VALUE)]
+  (let [client (nrepl/client transport Long/MAX_VALUE)
+        awaiting (atom {})]
     (println (repl-intro))
-    ;; We take 50ms to listen to any greeting messages, and display the value
-    ;; in the `:out` slot.
```
Comment (alexander-yakushev) [inline src/clojure/nrepl/cmdline.clj:99]: Ditto.

File: src/clojure/nrepl/cmdline.clj (hunk 5 above)
Comment (SevereOverfl0w) [reply src/clojure/nrepl/cmdline.clj:105]: Ah, I missed that subtly.  Cool.

**Comment by alexander-yakushev** [comment]
Awesome, thanks!

**Comment by SevereOverfl0w** [comment]
Thank you for merging!  I'm running this locally quite actively, I'll keep an eye out for any oddities.

**Pull request merged**